curated_transformers/layers/attention.py

```diff
@@ -173,23 +174,140 @@ class QkvMode(IntEnum):
     MERGED_SPLIT_AFTER = (2,)
 
 
-# https://www.tensorflow.org/text/tutorials/transformer#scaled_dot_product_attention
+class AttentionLinearBiases(Module):
+    """
+    ALiBi: Linear biases for attention (`Press et al., 2022`_).
+
+    .. _Press et al., 2022: https://arxiv.org/abs/2108.12409
+    """
+
+    slopes: Tensor
+
+    def __init__(self, *, num_attention_heads: int, is_causal: bool) -> None:
+        """
+        Construct an ALiBi module.
+
+        :param num_attention_heads:
+            Number of attention heads.
+        :param is_causal:
+            Use causal attention.
+        """
+        super().__init__()
+
+        self.is_causal = is_causal
+        slopes = self._calculate_slopes(num_attention_heads)
+        self.register_buffer("slopes", slopes, persistent=False)
+
+    def _calculate_slopes(self, num_attention_heads: int) -> Tensor:
+        """
+        Calculate the linear bias slopes for a given number
+        of attention heads.
+
+        :param num_attention_heads:
+            Number of attention heads.
+        :returns:
+            Slope tensor.
+
+            *Shape:* ``(1, heads, 1, 1)``
+
+        :meta private:
+        """
+        # Calculate values to the closest power of two and concat the remaining.
+        closest_pow_2 = 2 ** math.floor(math.log2(num_attention_heads))
+        ratio_base = 2.0 ** (-8.0 / closest_pow_2)
+        ratios = torch.pow(ratio_base, torch.arange(1, 1 + closest_pow_2))
+
+        # Start with for 2*closest_pow_2 and pick every other position
+        # to avoid previously picked values.
+        if closest_pow_2 < num_attention_heads:
+            new_ratio_base = 2.0 ** (-4.0 / closest_pow_2)
+            num_remaining_heads = min(
+                closest_pow_2, num_attention_heads - closest_pow_2
+            )
+            new_ratios = torch.pow(
+                new_ratio_base, torch.arange(1, 1 + 2 * num_remaining_heads, 2)
+            )
+            ratios = torch.cat([ratios, new_ratios])
```

Comment:
Could you explain why this is needed? From the paper, it seems that the slopes can be computed as:

```suggestion
        ratio = 2 ** (-8/num_attention_heads)
        ratios = ratio ** torch.arange(1, num_attention_heads + 1)
```